src/main/java/uk/gov/ons/ctp/response/casesvc/service/CaseService.java

```diff
@@ -167,6 +169,86 @@ public Case findCaseByIac(final String iac) throws CTPException {
     return cazes;
   }
 
+  /**
+   * Find the paginated cases for a partyId ordered by create date descending.
+   *
+   * @param partyId the partyId
+   * @param maxCasesPerSurvey the maximum number of cases per survey to return
+   * @return the cases for the partyId
+   */
+  public List<Case> findCasesByPartyIdLimitedPerSurvey(
+      final UUID partyId, boolean iac, int maxCasesPerSurvey) {
+    log.info("Entering findCasesByPartyIdLimitedPerSurvey");
+
+    List<Case> cazes =
+        limitCasesPerSurvey(
+            caseRepo.findByPartyIdOrderByCreatedDateTimeDesc(partyId), maxCasesPerSurvey);
+
+    if (iac) {
+      cazes.stream().forEach(c -> c.setIac(caseIacAuditService.findCaseIacByCasePK(c.getCasePK())));
+    }
+
+    return cazes;
+  }
+
+  /**
+   * Return only the first n instances of case per survey The query to this in the JPA is too
+   * complex and uses SQL constructs that JPQL does not support like Row Number and Over. Hence in
+   * code
+   *
+   * @param raw, the list of cases to be potentially limited
+   * @param maxCasesPerSurvey the maximum number of cases per survey to return
+   * @return the cases for the partyId
+   */
+  public List<Case> limitCasesPerSurvey(List<Case> raw, int maxCasesPerSurvey) {
+
+    log.with("maxCasesPerSurvey", maxCasesPerSurvey).info("Entering limitCasesPerSurvey");
+
+    List<Case> limitedResults = new ArrayList();
+    HashMap<UUID, Integer> surveyCounts = new HashMap<>();
+
+    raw.stream()
```

Comment:
I'm not convinced we need a new method for `tryAddCaseToLimitedResults` - this method is very small and this is the only method that will call TACTLR. I'd just write this as `for (Case case : raw) { ... logic ... }` rather than getting cute with streaming it.

I think the case for this is shown by the fact we have to keep passing in `limitedResults` and `surveyCounts` - it's just not a natural candidate for extraction to a method.

Reply:
Changing

Reply:
Changed